src/client.ts

```diff
@@ -2,14 +2,147 @@ import * as Models from './models'
+    super(error, response)
+    this.name = 'ApiNotAvailable'
+    Object.setPrototypeOf(this, ApiNotAvailable.prototype)
+  }
+}
+
+export class BadRequestApiError extends BaseApiError {
+  constructor(error: Models.ApiErrorUnion, response?: HttpOperationResponse) {
+    super(error, response)
+    this.name = 'BadRequestApiError'
+    Object.setPrototypeOf(this, BadRequestApiError.prototype)
+  }
+}
+
+export class UnauthorizedApiError extends BaseApiError {
+  constructor(error: Models.ApiErrorUnion, response?: HttpOperationResponse) {
+    super(error, response)
+    this.name = 'UnauthorizedApiError'
+    Object.setPrototypeOf(this, UnauthorizedApiError.prototype)
+  }
+}
+
+export class ForbiddenApiError extends BaseApiError {
+  constructor(error: Models.ApiErrorUnion, response?: HttpOperationResponse) {
+    super(error, response)
+    this.name = 'ForbiddenApiError'
+    Object.setPrototypeOf(this, ForbiddenApiError.prototype)
+  }
+}
+
+export class NotFoundApiError extends BaseApiError {
+  constructor(error: Models.ApiErrorUnion, response?: HttpOperationResponse) {
+    super(error, response)
+    this.name = 'NotFoundApiError'
+    Object.setPrototypeOf(this, NotFoundApiError.prototype)
+  }
+}
+
+export class InternalServerApiError extends BaseApiError {
+  constructor(error: Models.ApiErrorUnion, response?: HttpOperationResponse) {
+    super(error, response)
+    this.name = 'InternalServerApiError'
+    Object.setPrototypeOf(this, InternalServerApiError.prototype)
+  }
+}
+
+class ApiErrorHandlerPolicy extends BaseRequestPolicy {
+  public constructor(readonly _nextPolicy: RequestPolicy, readonly _options: RequestPolicyOptions) {
+    super(_nextPolicy, _options)
+  }
+
+  public async sendRequest(webResource: WebResource): Promise<HttpOperationResponse> {
+    try {
+      const result = await this._nextPolicy.sendRequest(webResource)
+      const body = result.parsedBody
+
+      if (body !== undefined && 'error' in body && 'message' in body) {
+        // the API responded with an expected error response
+        const exception = body as Models.ApiErrorUnion
+
+        switch (exception.error) {
+          case 'BadRequest':
+            throw new BadRequestApiError(exception, result)
+          case 'Unauthorized':
+            throw new UnauthorizedApiError(exception, result)
+          case 'Forbidden':
+            throw new ForbiddenApiError(exception, result)
+          case 'NotFound':
+            throw new NotFoundApiError(exception, result)
+          case 'InternalServerError':
+            throw new InternalServerApiError(exception, result)
+          default:
+            throw new UnexpectedApiError(exception, result)
+        }
+      } else if (result.status < 200 || result.status >= 300) {
+        // the API responded with an error status code, but not formed as an error - THIS SHOULD NEVER HAPPEN
+        throw new UnexpectedApiError(
```

Comment:
Can we rename this `UnexpectedApiError` (the one above should stay) and the `ApiNotAvailable` below?

ApiNotAvailable seems even more unexpected, and I would rather avoid the word all-together.

Maybe `MalformedApiError` instead of `UnexpectedApiError` since it's an error code without the correct format. 
As for `ApiNotAvailable`, this can be anything from a NetworkTimeout and NetworkNotAvailable (which fails instantly), to parsing errors from msRest. `UnhandledException` perhaps?

Anyhow, for the last 2 errors which are outside the L93 if, their constructors should receive the original error also so we can backtrace correctly.